tests/publiccloud/sles4sap.pm

```diff
@@ -27,15 +37,187 @@ sub upload_ha_sap_logs {
     }
 }
 
+=head2 wait_until_resources_started
+
+    wait_until_resources_started();
+
+Wait for HA cluster to be started.
+=cut
+sub wait_until_resources_started {
+    my ($self, %args) = @_;
+    my @cmds    = ('crm cluster wait_for_startup');
+    my $timeout = bmwqemu::scale_timeout($args{timeout} // 60);
+
+    # HANA cluster doesn't restart all resources after fencing in some configurations
+    if ($args{cluster_type} ne 'hana') {
+        # Some CRM options can only been added on recent versions
+        push @cmds, "$crm_mon_cmd | grep -iq no\\ inactive\\ resources" if is_sle '12-sp3+';
+    }
+
+    # Execute each comnmand to validate that the cluster is running
+    # This can takes time, so a loop is a good idea here
+    foreach my $cmd (@cmds) {
+        # Each command execution has its own timeout, so we need to reset the counter
+        my $starttime = time;
+
+        # Check for cluster/resources status and exit loop when needed
+        my $ret;
+        while ($ret = $self->{my_instance}->run_ssh_command(cmd => "sudo $cmd", rc_only => 1)) {
+            # Otherwise wait a while if timeout is not reached
+            if (time - $starttime < $timeout) {
+                sleep 5;
+            }
+            else {
+                # Cluster doesn't start properly, show state before dying
+                $self->run_cmd(cmd => $crm_mon_cmd);
+                die "Cluster/resources did not start within $timeout seconds (cmd='$cmd')";
+            }
+        }
+
+        # $ret need to be defined to ensure a correct exit code
+        if (!defined $ret) {
+            # Cluster doesn't start properly, show state before dying
+            $self->run_cmd(cmd => $crm_mon_cmd);
+            die 'Cluster/resources check did not exit properly';
+        }
+    }
+}
+
+=head2 run_cmd
+    run_cmd(cmd => 'command', [timeout => 60]);
+
+Runs a command C<cmd> via ssh in the given VM and log the output.
+All commands are executed through C<sudo>.
+=cut
+sub run_cmd {
+    my ($self, %args) = @_;
+    die('Argument <cmd> missing') unless ($args{cmd});
+    my $timeout = bmwqemu::scale_timeout($args{timeout} // 60);
+    my $title   = $args{title} // $args{cmd};
+    $title =~ s/[[:blank:]].+// if !defined $args{title};
+
+    my $out = $self->{my_instance}->run_ssh_command(cmd => "sudo $args{cmd}", timeout => $timeout);
+    record_info("$title output - $self->{my_instance}->{instance_id}", $out);
+}
+
+=head2 fence_node
+
+    fence_node([hostname => $hostname], [timeout => 60]);
+
+Fence a HA node using C<systemctl poweroff>.
+This command terminates when the node is back.
+=cut
+sub fence_node {
+    my ($self, %args) = @_;
+    my $timeout = bmwqemu::scale_timeout($args{timeout} // 60);
+    $args{hostname} //= $self->{my_instance}->{instance_id};
+
+    record_info('Fencing info', 'Fencing done with systemctl poweroff');
+    $self->{my_instance}->run_ssh_command(cmd => "sudo systemctl poweroff --force", timeout => 0);
```

Comment:
Why not `run_cmd` here?

Reply:
Simply because as I use `timeout => 0` there is no output, so nothing to log. And `run_cmd` is a wrapper on `run_ssh_command` to log output with `record`.

So not useful here.

Reply:
But you can pass a `timeout => 0` to `run_cmd()`, and skip the `record_info()` if the timeout is 0 in the function. Doesn't make much sense to me to have a wrapper function and use it only sometimes. Only L163 seems to be an exception IMO.

Reply:
Yes true, I will replace it and remove the current `record_info`, the one with the command should be enough.

> Only L163 seems to be an exception IMO.

A little bit more TBH, as there is 2 or 3 other lines like this one ;-)